sdk/src/ingredient.rs

```diff
@@ -841,36 +843,61 @@ impl Ingredient {
         };
 
         stream.rewind()?;
-        self.add_stream_internal(&format, stream)
+
+        if _sync {
+            self.add_stream_internal(&format, stream)
+        } else {
+            self.add_stream_internal_async(&format, stream).await
+        }
     }
 
     // Internal implementation to avoid code bloat.
+    #[async_generic()]
     fn add_stream_internal(mut self, format: &str, stream: &mut dyn CAIRead) -> Result<Self> {
         let mut validation_log = DetailedStatusTracker::new();
 
         // retrieve the manifest bytes from embedded, sidecar or remote and convert to store if found
-        let (result, manifest_bytes) = match load_jumbf_from_stream(format, stream) {
-            Ok(manifest_bytes) => {
-                (
-                    // generate a store from the buffer and then validate from the asset path
-                    Store::from_jumbf(&manifest_bytes, &mut validation_log)
-                        .and_then(|mut store| {
-                            // verify the store
-                            store.verify_from_stream(stream, format, &mut validation_log)?;
-                            Ok(store)
-                        })
-                        .map_err(|e| {
-                            // add a log entry for the error so we act like verify
-                            validation_log.log_silent(
-                                log_item!("asset", "error loading file", "Ingredient::from_file")
-                                    .set_error(&e),
-                            );
-                            e
-                        }),
-                    Some(manifest_bytes),
-                )
-            }
-            Err(err) => (Err(err), None),
+        let jumbf_stream = load_jumbf_from_stream(format, stream);
+
+        // We can't use functional combinators since we can't use async callbacks (https://github.com/rust-lang/rust/issues/62290)
+        let (result, manifest_bytes) = if let Ok(manifest_bytes) = jumbf_stream {
+            let jumbf_store = Store::from_jumbf(&manifest_bytes, &mut validation_log);
+            let result = if let Ok(mut store) = jumbf_store {
+                if _sync {
+                    match store.verify_from_stream(stream, format, &mut validation_log) {
+                        Ok(_) => Ok(store),
+                        Err(err) => Err(err),
+                    }
+                } else {
+                    match store
+                        .verify_from_stream_async(stream, format, &mut validation_log)
+                        .await
+                    {
+                        Ok(_) => Ok(store),
+                        Err(err) => Err(err),
+                    }
+                }
+            } else {
+                // This will always be Err in this situation
+                #[allow(clippy::unwrap_used)]
+                Err(jumbf_store.unwrap_err())
```

Comment:
There might be a nicer way to do this, however I wasn't sure how to handle this case without `match`, which I can't use since it creates a new function block inside the arms, which stops `async_generic` from working properly (would have to conditionally provide a sync or async function wrapper).

Reply:
it could be cleaner but it is ok